Tests/SwiftProtobufTests/Test_AsyncMessageSequence.swift

```diff
@@ -0,0 +1,223 @@
+// Tests/SwiftProtobufTests/Test_AsyncMessageSequence.swift - 
+//
+// Copyright (c) 2023 Apple Inc. and the project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See LICENSE.txt for license information:
+// https://github.com/apple/swift-protobuf/blob/main/LICENSE.txt
+//
+// -----------------------------------------------------------------------------
+///
+/// Tests the decoding of binary-delimited message streams, ensuring various invalid stream scenarios are
+/// handled gracefully.
+///
+// -----------------------------------------------------------------------------
+
+import Foundation
+import XCTest
+import SwiftProtobuf
+
+final class Test_AsyncMessageSequence: XCTestCase {
+  
+  // Decode a valid binary delimited stream
+  func testValidSequence() async throws {
+    let expected: [Int32] = Array(1...5)
+    var messages = [SwiftProtoTesting_TestAllTypes]()
+    for messageNumber in expected {
+      let message = SwiftProtoTesting_TestAllTypes.with {
+        $0.optionalInt32 = messageNumber
+      }
+      messages.append(message)
+    }
+    let serialized = try serializedMessageData(messages: messages)
+    let asyncBytes = asyncByteStream(bytes: serialized)
+    
+    // Recreate the original array
+    let decoded = asyncBytes.binaryDelimitedMessages(of: SwiftProtoTesting_TestAllTypes.self)
```

Comment:
> I'm not thrilled about extending other people's types in this way, though I agree this reads pretty nicely.

This is totally normal with `AsyncSequence`s and a pattern that `swift-async-algorithms` and the standard library follows e.g. `.map {}` on `AsyncSequence comes from the standard library.

Reply:
Adding to this. If this is not an extension the chainability gets completely broken. You want to be able to write this

```
rootAsyncSequence.map { ... }.binaryProtobufDelimitedMessages().map { ... }.filter { ... }
```

Reply:
I am with @tbkka on this, I don't think that publicly extending such a fundamental type is appropriate. Independent of how this example reads.

Reply:
> I am with @tbkka on this, I don't think that publicly extending such a fundamental type is appropriate. 

This is how `AsyncSequence` chaining has been envisioned and how we have spelled this in both the standard library and `swift-async-algorithms`. This is very much the reactive chaining approach and if we would not add that extension the usability of this addition will degrade significantly.

Reply:
In the case of the standard library (which I believe introduces `AsyncSequence` to begin with) and the async algorithms package, the _primary subject_ under consideration is the sequence and the way it composes. The methods that those modules introduce provide abstract and generic functionality over the sequence.

In our case, the subject of this module is not the sequence, we do not define or own the type, and neither is this extensions abstract and generic.

I fully appreciate the syntactic value of a method provided by an extension, and I love introducing nice composability. But I don't think that it is fair to say polluting the namespace of the members of a public type is what anyone has envisioned. I also don't think usability will degrade significantly--at most it will degrade slightly. And I also think it's not our job here to fix this particular defect in the ecosystem. It would be more proper for a new abstract chunking combinator (that could handle the varint delimiters and reduce the bytes) or even something like the F# pipe operator, which is specifically amazing at fixing the ordering of these things in syntax.

As a thought exercise, if we were to later add this functionality for decoding JSON messages, do you think anyone would object to putting a method on `String` called `jsonProtobufDelimitedMessages()`?

Reply:
I'm okay with this as long as we adopt the more specific name.  Resolving this for now so we can land this PR.